### DISCUSSION TITLE
Warnings appears during training related to phonemizer on arabic language

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#313] [Archived]
<USER>>>> khalilrhouma<br/> <DATE> [December 4, 2020, 9:48am]<br/> <br/> **Working branch** : dev<br/> <br/> I am trying to fine-tune Tacotron-DDC-130k on LJSpeech on the Arabic<br/> dataset using the phonemizer. Arabic is supported under `ar` tag. *slash* <br/> During the training, these warnings didn't stop for the entire epoch:<br/> <br/> [WARNING] extra phones may appear in the 'ar' phoneset<br/> [WARNING] language switch flags have been removed (applying 'remove-flags' policy)<br/> [WARNING] 1 utterances containing language switches on lines 1<br/> [WARNING] extra phones may appear in the 'ar' phoneset<br/> [WARNING] language switch flags have been removed (applying 'remove-flags' policy)<br/> [WARNING] 1 utterances containing language switches on lines 1<br/> [WARNING] extra phones may appear in the 'ar' phoneset<br/> [WARNING] language switch flags have been removed (applying 'remove-flags' policy)<br/> [WARNING] 1 utterances containing language switches on lines 1<br/> [WARNING] extra phones may appear in the 'ar' phoneset<br/> [WARNING] language switch flags have been removed (applying 'remove-flags' policy)<br/> [WARNING] 1 utterances containing language switches on lines 1<br/> ............<br/> <br/> Where the problem came from? I mean is there anything to change before<br/> training except these parameters:<br/> <br/> text_cleaner: phoneme_cleaners<br/> use_phonemes: true<br/> phoneme_language: ar<br/> <br/> I think it may be related to text cleaning Any help? *slash* <br/> Thank you<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/warnings-appears-during-training-related-to-phonemizer-on-arabic-language]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> khalilrhouma<br/> <DATE> [December 4, 2020, 11:10am]<br/> <br/> *slash* <br/> yes I posted here because I didn't get a replay for the issue , I close<br/> it now as you told me. *slash* <br/> for the dataset it is about 3h, this is a sample of diacritic Arabic<br/> which is supported by the phonemizer:<br/> <br/> > إِضَافَةً إِلَى عَرْضِ أَنْوَاعٍ عَدِيدَةٍ مِنَ الْأَثَاثِ<br/> > الْمُزَخْرَفِ وَالصَّنَادِيقِ وَالْأَوَانِي وَالْأَدَوَاتِ -<br/> > وَالْأَغْطِيَةِ وَاللَّوْحَاتِ الْفَنِّيَّةِ الْعَاكِسَةِ<br/> > لِتِقْنِيَّاتٍ مُدْهِشَةٍ - تُجَسِّدُ مَهَارَاتٍ حِرَفِيَّةً<br/> > عَرَفَتْهَا الْأَنْدُلُسْ<br/> <br/> we tested the phonemizer on the Arabic sentence directly and the<br/> phonemizer didn't output any warning. This is the phonemizer output:<br/> <br/> > ʔidˤa.ːfatan ʔilaː ʕardˤi. ʔanwaːʕin ʕadiːdatin mina alʔaθaːθi<br/> > almuzaχɹafi was̪ːnaːdiːqi walʔauaːniː walʔadauaːti walʔaɣt̪i.ːati<br/> > wallwħaːti alfanniːti alʕaːkisati litiqniːaːtin mudhiʃatin tudʒassdu<br/> > mahaːɹaːtin ħiɹafiːtan ʕaɹafathaː alʔandulus<br/> <br/> which is correct without any error or warnings, that's why I am thinking<br/> it is related to text cleaning form mozilla tts<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> othiele<br/> <DATE> [December 4, 2020, 12:51pm]<br/> <br/> Put in some logging where TTS uses the phonemizer and find out what<br/> utterances cause the messages. For us we then knew what we had to<br/> change.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [December 4, 2020, 2:21pm]<br/> <br/> I'd suggest to run phonemizer over all of your dataset, list unique<br/> phones and compare it with the phoneme set under symbols.py. If there<br/> are extras, you can add them there and send a PR for us to update it.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> khalilrhouma<br/> <DATE> [December 4, 2020, 4:59pm]<br/> <br/> *slash* <br/> We tested the whole dataset with phonemizer only, it gives it's output<br/> as expected without any warnings. *slash* <br/> Actually, the problem was related to phonemizer_cleaner , and exactly<br/> when the ASCII conversion happens: `text = convert_to_ascii(text)` *slash* <br/> Phonemizer accepts the Arabic letters as it is , and when you make that<br/> conversion it does change the letters completely to an other format<br/> which is not Arabic. *slash* <br/> So, for now we are using `basic_cleaners` in place of `phoneme_cleaners`<br/> and the warnings gone (until this moment) *slash* <br/> Is there any method to disable cleaners ?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [December 4, 2020, 5:08pm]<br/> <br/> you can pass a empty function which only returns the value as it is.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> mrthorstenm<br/> <DATE> [December 4, 2020, 6:42pm]<br/> <br/> .<br/> <br/> Didn't have time to answer more early. *slash* <br/> Here are some info what we did in german phoneme cleaning.<br/> <br/> We appended following lines in the end of 'TTS/TTS/tts/utils/text'<br/> <br/> def german_phoneme_cleaners(text):<br/> return GermanTransliterate(replace={';': ',', ':': ' '}, sep_abbreviation=' -- ').transliterate(text)<br/> <br/> Maybe that's the place where you could make your own function returning<br/> the original value. *slash* <br/> Reference your method name in your config.json as we did:<br/> <br/> 'text_cleaner': 'german_phoneme_cleaners',<br/> <br/> We had some discussion on phoneme topic in another thread, maybe this is<br/> helpful for you. *slash* <br/> <br/> ![](../../user_avatar/discourse.mozilla.org/mrthorstenm/40/31471_2.png){.avatar<br/> tts](https://discourse.mozilla.org/t/contributing-my-german-voice-for-tts/48150/141)<br/> drop-close='true'<br/> .box}<br/> <br/> > Hello. It's time for another short update. We're currently preparing a<br/> > new training run and figured out some 'issues' with phoneme handling<br/> > with mixed english/german wording. This warning occurs quite often:<br/> > *slash* [WARNING *slash* ] fount 1 utterances containing language switches on lines 1<br/> > *slash* [WARNING *slash* ] extra phones may appear in the *slash* 'de *slash* ' phoneset *slash* [WARNING *slash* ]<br/> > language switch flags have been kept (applying *slash* 'keep-flags *slash* ' policy)<br/> > We analyzed where these warnings come from and found out that our<br/> > dataset (metadata.c...<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> khalilrhouma<br/> <DATE> [December 5, 2020, 7:15am]<br/> <br/> Thank you all for your support and your time. *slash* <br/> said, we just need a<br/> function that returns the input as it is. The `basic_cleaners` function<br/> seems to do that properly, and we check that it gives the input as it is<br/> without issues. *slash* <br/> So thanks again, I really appreciate your work in Mozilla TTS<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> Pak<br/> <DATE> [February 10, 2021, 12:34pm]<br/> <br/> > `basic_cleaners`<br/> <br/> Hello, *slash* <br/> Can you share your pipeline for training arabic? have own DB, dont need<br/> data. *slash* <br/> Thanks you.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [February 10, 2021, 1:24pm]<br/> <br/> so did it work? How you got your Arabic model?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> Pak<br/> <DATE> [February 11, 2021, 6:03am]<br/> <br/> I install espeak-ng to get 'ar' support, but get<br/> 'xcb_connection_has_error() returned true' so fix by make 'unset<br/> DISPLAY' now train. Will report result.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [February 11, 2021, 1:12pm]<br/> <br/> Google is your friend. Not related to TTS.<br/> <br/> [Archived Post]<br/>

**Comment by cod3r0k** [comment]
Hi , do you have any update about preliminaries steps for Arabic language?
I mean, do I need to set `guut` as phonemizer with `ar` language? and then use `phoneme_cleaners` ?  
Also do we need to extends punctuations as below?
```
tokenizer, config = TTSTokenizer.init_from_config(config)
tokenizer.characters.punctuations = "!'(),-.:;? ؟،؛ ‌ \"“”«»"
```  @JRMeyer @all
